### PR TITLE
More lock contention reductions

### PR DESCRIPTION
**Pull request by pettyjamesm**
Continues lock contention reductions from https://github.com/prestodb/presto/pull/15477

Seven commits:
- Cache `isBlocked` future in various operators to avoid repeatedly contending for locks and potentially registering new callbacks once those operators are blocked
- Add unsynchronized fast-paths in `ExchangeClient` to avoid unnecessary contention
- Refactor `ExchangeClient#pollPage` to avoid redundant checks and synchronizations
- Use `AtomicIntegerFieldUpdater` for local exchange `PageReference`
- Reduce synchronization in `LocalExchangeMemoryManager` by making unblocking an edge-triggered event that only a single caller initiates.
- Add unsynchronized common-case fast paths in `LocalExchangeSource`
- Remove synchronized for final field read in `MemoryPool#getMaxBytes()`

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSource.java
Comment (aweisberg) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSource.java:19]: Not controversial either. But do the changes around managing the futre have an impact? You are removing some volatile reads yes, but you are also adding conditional logic under the lock.

**Review by aweisberg — CHANGES_REQUESTED** [review]
Most of these changes LGTM modulo some correctness questions.

More generally I can only vouch that these changes look like they shouldn't break anything, but some of them do add a bit of weight.

Were these changes made on inspection? Profile guided? Are we going to do a bunch of these and accept that only some will have an impact we care about?

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java:95]: All accesses of this field (and operators in general) is done while holding the driver lock in a single threaded manner, so unsafe publication shouldn't be a concern. Even if that weren't true, the only risk would be somehow seeing `null` in this field because all of the `ListenableFuture<?>` instances returned here are amenable to being called multiple times- just at the cost of synchronizing into a critical section.

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java:456]: Indeed it does, I think the best route here is going to be to clear the blocked callers list while holding the lock and notifying them after releasing it.

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java:428]: Fair point, I'll remove the pre-checks.

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java:463]: Yep, even adding a single page to the buffer immediately wakes up any and all waiters so they have to be tolerant to that possibility- still, not a desirable thing to do but addressing that is non-trivial complexity wise.

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSource.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSource.java:19]: True, but this will be a "shallow check" in that we're just comparing the reference value and not accessing the target of the reference. As such, one would expect this to amount to a very-local CPU cache entry because we've already taken ownership of the object monitor.

Ultimately it's not obvious whether to unconditionally save and clear the listener without branching is better here. It could well be that the CPU wouldn't issue a store operation if the value was null and null was being set.

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java:92]: I think probably worse. I'm not sure how one would accomodate the _slight_ variations in the behaviors of where the `isBlocked` trigger comes from without cluttering the `Operator` interface.

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java:95]: I don't think there is a guarantee that the ListenableFuture implementation is constructed in such a way that after the constructor completes the fields are visible? I did look at all the callers and yeah it looks like it is all from the driver loop I just didn't have as high a confidence as you that I wasn't missing something.

LGTM

File: presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java]: In the common case pages will not be empty since one was just added so there is no reason to release the lock before acquiring the list of callers?

How common will it be that the list of blocked callers is non-empty? I imagine it depends a lot on the query and whether the producer or the consumer is faster.

Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java]: > In the common case pages will not be empty since one was just added so there is no reason to release the lock before acquiring the list of callers?

You're correct. Originally I restructured it this way to avoid holding the lock for the duration of the notifications being enqueued to the background executor (`ScheduledThreadPoolExecutorService` have a priority queue as their run queue even for non-scheduled runnables, so enqueues are more expensive than you might expect). I've updated the commit to just duplicate a little snippet of the `notifyBlockedCallers()` method with the synchronization in the right place.

> How common will it be that the list of blocked callers is non-empty? I imagine it depends a lot on the query and whether the producer or the consumer is faster.

It should depend on the relative pace of consumers + producers, but I think it's probably very common for the exchange network throughput is slower than whatever consumes from the other side of the exchange. As an aside, before the operator `ListenableFuture<?> isBlocked` caching change each driver would enqueue at least 3 separate listeners because of the way that the Driver loop discards the future on the first pass.

**Review by aweisberg — APPROVED** [review]
I talked with @pettyjamesm and these changes were profile guided and with all the PRs combined there was a several X reduction in time spent contending on locks for his test scenario.

Let's wait for the PrestoSQL review to complete before merging just in case they find anything.

**Comment by pettyjamesm** [comment]
@aweisberg based on the review feedback in https://github.com/trinodb/trino/pull/6260 I've made the following changes here:
- added https://github.com/prestodb/presto/pull/15509/commits/5ca13df409e24e13753aae8209625858afdf00a5 
- changed https://github.com/prestodb/presto/pull/15509/commits/d28a0b16cd4377c9196155f95a1009000c5043d9 to move other potentially contended lock accesses inside of the `if (!pages.isEmpty())` block

**Comment by aweisberg** [comment]
Why we would call addPages if there are no pages is ??? but +1 to both.

**Comment by pettyjamesm** [comment]
> Why we would call addPages if there are no pages is ??? (and acquire a lock to do it) but +1 to both.

I believe it's necessary for the average response size metrics to reflect empty responses having been received. That's certainly what was happening before and [this comment](https://github.com/prestodb/presto/blob/6ff3052e9df1a04613157746e7e779fd54c0c2a1/presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java#L313) implies that it's important to preserve.

**Comment by wenleix** [comment]
Merged #15509. Thanks for the contribution!

**Pull request merged**
